src/sources/splunk_hec/mod.rs

```diff
@@ -191,32 +234,47 @@ impl SplunkSource {
                     let mut out = out
                         .clone()
                         .sink_map_err(|_| Rejection::from(ApiError::ServerShutdown));
+                    let idx_ack = idx_ack.clone();
                     emit!(&HttpBytesReceived {
                         byte_size: body.len(),
                         http_path: path.as_str(),
                         protocol,
                     });
+
                     async move {
+                        if idx_ack.is_some() && channel.is_none() {
+                            return Err(Rejection::from(ApiError::MissingChannel));
+                        }
+
                         let reader: Box<dyn Read + Send> = if gzip {
                             Box::new(MultiGzDecoder::new(body.reader()))
                         } else {
                             Box::new(body.reader())
                         };
 
-                        let events = stream::iter(EventIterator::new(
+                        let (batch, receiver) =
+                            BatchNotifier::maybe_new_with_receiver(idx_ack.is_some());
+                        let mut events = stream::iter(EventIterator::new(
                             Deserializer::from_reader(reader).into_iter::<JsonValue>(),
-                            channel,
+                            channel.clone(),
                             remote,
                             xff,
+                            batch,
                         ));
 
-                        // `fn send_all` can be used once https://github.com/rust-lang/futures-rs/issues/2402
-                        // is resolved.
-                        let res = events.forward(&mut out).await;
+                        let maybe_ack_id = match (idx_ack, receiver, channel) {
+                            (Some(idx_ack), Some(receiver), Some(channel_id)) => {
+                                match idx_ack.get_ack_id_from_channel(channel_id, receiver).await {
+                                    Ok(ack_id) => Some(ack_id),
+                                    Err(rej) => return Err(rej),
+                                }
```

Comment:
Ah yes, I'd tried the above but ran into an odd error (below) doing it in `raw_service`. It works in `event_service`, which further confuses me. What might be happening here?
```
error[E0283]: type annotations needed
   --> src/sources/splunk_hec/mod.rs:299:14
    |
299 |             .and_then(
    |              ^^^^^^^^ cannot infer type
    |
    = note: cannot satisfy `_: reject::sealed::CombineRejection<Rejection>`
help: consider specifying the type argument in the method call
    |
299 |             .and_then::<F>(
    |                      +++++
```

Reply:
I'm not seeing where this error is happening for you. In general, though, this will be because there are multiple contexts inferring types, and rustc can't figure out exactly which one is required. So, the type returned by the closure is indeterminate, and the caller of the closure also doesn't have an inferred type.

Reply:
Ah yes, I think I'd confused the type inference with multiple `?` operators. Updated in 8d98858